### ISSUE TITLE
ModernFix causing frequent server crashes

### ISSUE THREAD

**Issue by SageofTurtles** [embeddedt/ModernFix#446]
### Bug Description

Our server has been experiencing repeated crashes over the last week, which (after extensive testing) seem to be resolved only by removing ModernFix from our modpack.  Possibly an incompatibility with another mod, but I'm having a lot of difficulty reproducing the issue, so I can't give too much information for this.  There are a few different errors given in the crash reports/logs when this happens, but they all seem very similar, which leads me to believe the attached crash reports were all triggered by the same issue.  We've done numerous tests on the one repeatable instance we were able to isolate (despite not knowing what was causing the crashes even in that instance) using versions 5.19.0, 5.18.10, 5.18.6, and 5.18.0 of your mod for Fabric 1.20.1, but every tested version seems to cause this same error.  Only by removing ModernFix were we able to perform our test case without a server crash.

### Reproduction Steps

I genuinely have no clue.  Our test case was one of our players placing a row of soul sand in a certain area of his base in the world.  We caught the crashes happening while he was doing this, and saved a backup of the world to perform troubleshooting on.  Once we isolated and removed ModernFix, the crashes no longer occurred, and we removed the mod from our main server.  This issue hasn't occurred in the 12 hours since that time, so I'm fairly confident that fixed it.  But again, I could never find what exactly was triggering this to happen, only that it occurred in one specific case and did not occur when the mod was removed.

### Log File

Here are the crash reports for our server from the last week that seem to be related to this issue, as best I am able to determine.  The folder contains 32 of these crash reports.  I'm happy to provide any additional details I possibly can.

[Crash Reports.zip](https://github.com/user-attachments/files/16401157/Crash.Reports.zip)

**Comment by embeddedt** [comment]
Have you tried to reproduce this without C2ME? This looks like another mod changing entity chunk statuses on the wrong thread (which would lead to random exceptions from fastutil). I am not sure that ModernFix is actually causing the issue; it might just speed up some server process enough to make the crash happen more consistently. I also tried debugging locally with just ModernFix, and could not recreate a situation in which the crash would be possible.

**Comment by SageofTurtles** [comment]
C2ME was one of the first mods I tried removing, and the issue still occurred without it.  I do recognize the possibility that this is actually caused by something else, ModernFix was just the only catalyst for this type of crash that I could find by trail-and-error.

**Comment by embeddedt** [comment]
What happens if `mixin.bugfix.paper_chunk_patches` is disabled in the ModernFix config?

**Comment by VanillaChan6571** [comment]
Can Confirm. Just crashes/hangs for no reason and basically never recovers. It seems to dislike datapacks. If you get the message below.

**[ModernFix integrated server watchdog/ERROR]: A single server tick has taken 40001, more than 40000 milliseconds**

Then your client is softlocked and needs to Alt + F4.

**Comment by FireDragon91245** [comment]
Also having the same issue on 1.19.2 Forge

Sometimes on World gen or normal server Oprations server crashes

client same modpack seems to work fine

**Comment by Numilani** [comment]
Also experiencing this issue, albeit I've done little-to-no testing to determine what mods seem to be clashing or what settings might fix it. 

Was able to consistently reproduce the problem, though - flying over a specific section of terrain would apparently cause worldgen to lag out, the server watchdog would hit 40001 ms and crash, the world file wouldn't save, then loading the same file up would cause the same problem. Managed to crash the world three times in a row with the same problem, then tried removing ModernFix - no crashes.

Instance details below, excluding a modlist (I can pastebin that upon request):
```
Game Version: 1.20.1
Loader: Forge
Loader Version: 47.3.11
ModernFix Version: 5.19.4+mc1.20.1
```

Upon request, I'm happy to provide any information I can - I'd love to see this resolved, the performance of my pack is notably suffering without ModernFix.

**Comment by embeddedt** [comment]
Can you provide the log file when the watchdog kills it? That sounds like a different class of issue from what others here are experiencing, and probably easier to resolve.

**Comment by Numilani** [comment]
Sure, here's one example:

[2024-10-13-3.log.gz](https://github.com/user-attachments/files/17369321/2024-10-13-3.log.gz)

**Comment by vizthex123** [comment]
Got the same server tick message after running `/reload` on my 1.18.2 pack. I edited a KubeJS and CraftTweaker script, and didn't want to reload the entire game.
[latest.log](https://github.com/user-attachments/files/17605116/latest.log)

**Comment by Minebrah** [comment]
I got the same problem on the latest release for Forge 1.20.1.

**Comment by embeddedt** [comment]
@Numilani Your issue appears to be a deadlock caused by In Control.

```
"Server thread" prio=4 Id=141 TIMED_WAITING on java.lang.String@2e2e6540
	at java.base@21.0.4/jdk.internal.misc.Unsafe.park(Native Method)
	-  waiting on java.lang.String@2e2e6540
	at java.base@21.0.4/java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:269)
	at TRANSFORMER/minecraft@1.20.1/net.minecraft.util.thread.BlockableEventLoop.m_5667_(BlockableEventLoop.java:521)
	at TRANSFORMER/minecraft@1.20.1/net.minecraft.util.thread.BlockableEventLoop.m_18701_(BlockableEventLoop.java:141)
	at TRANSFORMER/minecraft@1.20.1/net.minecraft.server.level.ServerChunkCache.getChunkBlocking(ServerChunkCache.java:1768)
	at TRANSFORMER/minecraft@1.20.1/net.minecraft.server.level.ServerChunkCache.m_7587_(ServerChunkCache.java:1664)
	at TRANSFORMER/minecraft@1.20.1/net.minecraft.world.level.Level.m_6522_(Level.java:187)
	at TRANSFORMER/minecraft@1.20.1/net.minecraft.world.level.Level.m_6325_(Level.java:4051)
	at TRANSFORMER/minecraft@1.20.1/net.minecraft.world.level.Level.m_8055_(Level.java:4540)
	at TRANSFORMER/minecraft@1.20.1/net.minecraft.world.level.LevelReader.m_46861_(LevelReader.java:97)
	at TRANSFORMER/incontrol@1.20-9.2.7/mcjty.incontrol.rules.support.GenericRuleEvaluator.lambda$addSeeSkyCheck$35(GenericRuleEvaluator.java:394)
	at TRANSFORMER/incontrol@1.20-9.2.7/mcjty.incontrol.rules.support.GenericRuleEvaluator$$Lambda/0x0000000083b10f50.apply(Unknown Source)
	at TRANSFORMER/incontrol@1.20-9.2.7/mcjty.incontrol.rules.support.GenericRuleEvaluator.match(GenericRuleEvaluator.java:1236)
	at TRANSFORMER/incontrol@1.20-9.2.7/mcjty.incontrol.rules.SpawnRule.match(SpawnRule.java:427)
	at TRANSFORMER/incontrol@1.20-9.2.7/mcjty.incontrol.ForgeEventHandlers.onEntityJoinWorld(ForgeEventHandlers.java:83)
	at TRANSFORMER/incontrol@1.20-9.2.7/mcjty.incontrol.__ForgeEventHandlers_onEntityJoinWorld_EntityJoinLevelEvent.invoke(.dynamic)
```

**Comment by Veynam** [comment]
Just here to pitch in, having the EXACT same problem with latest versions of ModernFix as everyone else here

**Comment by gabepikamole** [comment]
same kind of issue :(

**Comment by n24486838** [comment]
[crash-2024-11-18_02.57.12-server.txt](https://github.com/user-attachments/files/17791997/crash-2024-11-18_02.57.12-server.txt)
Minecraft : 1.20.1 (forge 47.3.0)

Hello, I also encountered a very similar problem.
Recently, when I was playing a server with my friends, it would crash when loading into a certain section.

Looking at the crash report it seems to be a problem with ModernFix, can you help me figure out what went wrong?
(The above content is translated using Google, please forgive me for any semantic errors)

Thanks !

**Comment by johalun** [comment]
I'm also getting this frequently when using `/locate` or chunky.

**Comment by MrBOOMberMAN** [comment]
exact same problem here pls fix this

**Comment by Kimercal** [comment]
same issue 
[crash-2024-12-03_21.43.17-server.txt](https://github.com/user-attachments/files/18002377/crash-2024-12-03_21.43.17-server.txt)

**Comment by Cazadorsniper** [comment]
I began to have this issue as well, but solved it after removing an Incontrol rule that I had.

**Comment by ZenoGrave** [comment]
New to making modpacks and stuff. I'm thinking I might have the same or similar issue to these. Not the first time for me on my server that it's done this. I feel like my root might be some of the mods that add alot of entities, but honestly trying to learn as much as I can...
edit: the first time was issues with Nyf's Spiders mod and was resolved after removing it
[crash-2024-12-15_23.12.09-server.txt](https://github.com/user-attachments/files/18145367/crash-2024-12-15_23.12.09-server.txt)

**Comment by johalun** [comment]
For me this was the symptom of a problematic structure, causing MC to hang during chunk generation. Without ModernFix MC would still hang, but just no watchdog to report it.

**Comment by baronpalmer** [comment]
The mod "Nyf's Spiders" has been causing this to happen to me, if you have it installed try removing it
![image](https://github.com/user-attachments/assets/ad1778f1-dfa9-41a1-9eec-093dcbe13d6b)

**Comment by embeddedt** [comment]
Does anyone in this thread still experience the *original* issue (not some of the ones posted later, which are from unrelated mods) on 1.20.1? I can try to add some debugging logic to detect the cause, but it needs to be tested by someone who can reliably trigger the problem.

**Comment by gabepikamole** [comment]
last comment was made just so i could edit after troubleshooting. i used chatgpt (yeah, i know...) to get pointers, and the solution to my problem THIS time was disabling async chunk loading/worldgen from c2me. i got to this solution after dropping the server dump text modern fix gives us in there and waiting for it to somewhat digest the info, and it actually got it right.

**Comment by CreeperEntertain** [comment]
> The mod "Nyf's Spiders" has been causing this to happen to me, if you have it installed try removing it ![image](https://github.com/user-attachments/assets/ad1778f1-dfa9-41a1-9eec-093dcbe13d6b)

It's comically bad how bad performance with Nyf's Spiders is on modded servers. I thought about bringing this up to the dev, but previous replies to comments about the same issue were replied to with an attitude of "just don't play with it, then."

**Comment by baronpalmer** [comment]
> > The mod "Nyf's Spiders" has been causing this to happen to me, if you have it installed try removing it ![image](https://github.com/user-attachments/assets/ad1778f1-dfa9-41a1-9eec-093dcbe13d6b)
> 
> It's comically bad how bad performance with Nyf's Spiders is on modded servers. I thought about bringing this up to the dev, but previous replies to comments about the same issue were replied to with an attitude of "just don't play with it, then."

I had the same experience, removed all of his mods from my pack afterwards.

**Comment by CreeperEntertain** [comment]
> I had the same experience, removed all of his mods from my pack afterwards.

It's only the Random Village Names mod. You can safely keep the other ones in in my experience.

**Comment by baronpalmer** [comment]
> > I had the same experience, removed all of his mods from my pack afterwards.
> 
> It's only the Random Village Names mod. You can safely keep the other ones in in my experience.

I have never used the Random Village Names mod in my life, clearly we have different experiences.
Just want to say there's a fork of Nyf's Spiders called "Nyfs Spiders Au Naturel Edit" that works great

**Comment by PleahMaCaka** [comment]
I had the same issue while playing [FMTT(From Magic To Tech)](https://www.curseforge.com/minecraft/modpacks/fmtt). Removing 'nyfsspiders' solved it for me.

**Comment by LokeYourC3PH** [comment]
> Does anyone in this thread still experience the _original_ issue (not some of the ones posted later, which are from unrelated mods) on 1.20.1? I can try to add some debugging logic to detect the cause, but it needs to be tested by someone who can reliably trigger the problem.

I have a singleplayer world that faces this right now, and I can reliably trigger it since my world is currently in an unrecoverable state because of this.

**Comment by LokeYourC3PH** [comment]
> > Does anyone in this thread still experience the _original_ issue (not some of the ones posted later, which are from unrelated mods) on 1.20.1? I can try to add some debugging logic to detect the cause, but it needs to be tested by someone who can reliably trigger the problem.
> 
> I have a singleplayer world that faces this right now, and I can reliably trigger it since my world is currently in an unrecoverable state because of this.

To anyone else having this problem, and removing "Nyf's Spiders" not fixing it: It's [Starlight](https://modrinth.com/mod/starlight-forge) that causes this. Because, my world was still broken, and the other world (a new one I made), eventually ended up failing to load as well beyond a certain amount of chunks loaded after traveling far. Eventually, I had checked online for solutions to no ends, until someone (can't find the link anymore) had mentioned they removed Starlight, which fixed it.

And lo and behold: It did. In a modpack as huge as mine with 600+ mods, Starlight ended up being the one causing massive issues in many other ways I didn't even realize yet. It's sad, but really, I notice barely any difference without it.

So to those who face the same problem as me: Try removing Starlight.

**Comment by Galanskiy** [comment]
Dear community, maybe someone can help or point me in the right direction.

I created my own modpack for Minecraft 1.18.2 Forge (partially forked from DawnCraft, but mostly customized for my own needs). We were playing with friends over LAN (without setting up a dedicated server) and ran into a problem where, at some point, the server completely freezes or stops for various reasons:

- a large number of entities on a mob farm

- multiple players loading chunks simultaneously in different directions

- chests and storage systems at the base

- possibly frequent teleportation to each other (sometimes from different dimensions)

- or a combination of all of the above

Sometimes, when the server starts freezing, it just lags, then after a while rolls back and continues working. However, in some cases it results in a complete freeze / crash. A crash report is not always generated, but I managed to collect some logs.

I tried to analyze the issue using ChatGPT, providing logs and the mod list. It didn’t point to anything definitive, but ModernFix was mentioned as a possible suspect. I haven’t yet started manually disabling mods one by one and testing (since this is time-consuming).

ModernFix version: modernfix-forge-5.18.0+mc1.18.2.jar

Also mentioned mods (that may cause issue):

- Fast Async World Save

- Feature NBT Deadlock Be Gone

- ServerCore

- It Shall Not Tick

After researching the issue myself, searching for “Server thread stalled / tick timeout” and ModernFix, I came across this GitHub Issues page.

Of course, the issue could be caused by another mod or a conflict between multiple mods, but the problems described on this Issues page are very similar to what we experienced during gameplay.

P.S. I am attaching the logs I managed to collect while trying to identify the problem. If anyone has ideas or suggestions, I would really appreciate the help!

[crash-2026-01-15_23.51.29-client.txt](https://github.com/user-attachments/files/24995995/crash-2026-01-15_23.51.29-client.txt)
[crash-2026-01-15_23.51.27-server.txt](https://github.com/user-attachments/files/24995997/crash-2026-01-15_23.51.27-server.txt)
[log.txt](https://github.com/user-attachments/files/24995996/log.txt)

**Comment by Flimsy-Fox** [comment]
> Dear community, maybe someone can help or point me in the right direction.
> 
> I created my own modpack for Minecraft 1.18.2 Forge (partially forked from DawnCraft, but mostly customized for my own needs). We were playing with friends over LAN (without setting up a dedicated server) and ran into a problem where, at some point, the server completely freezes or stops for various reasons:
> 
> * a large number of entities on a mob farm
> * multiple players loading chunks simultaneously in different directions
> * chests and storage systems at the base
> * possibly frequent teleportation to each other (sometimes from different dimensions)
> * or a combination of all of the above
> 
> Sometimes, when the server starts freezing, it just lags, then after a while rolls back and continues working. However, in some cases it results in a complete freeze / crash. A crash report is not always generated, but I managed to collect some logs.
> 
> I tried to analyze the issue using ChatGPT, providing logs and the mod list. It didn’t point to anything definitive, but ModernFix was mentioned as a possible suspect. I haven’t yet started manually disabling mods one by one and testing (since this is time-consuming).
> 
> ModernFix version: modernfix-forge-5.18.0+mc1.18.2.jar
> 
> Also mentioned mods (that may cause issue):
> 
> * Fast Async World Save
> * Feature NBT Deadlock Be Gone
> * ServerCore
> * It Shall Not Tick
> 
> After researching the issue myself, searching for “Server thread stalled / tick timeout” and ModernFix, I came across this GitHub Issues page.
> 
> Of course, the issue could be caused by another mod or a conflict between multiple mods, but the problems described on this Issues page are very similar to what we experienced during gameplay.
> 
> P.S. I am attaching the logs I managed to collect while trying to identify the problem. If anyone has ideas or suggestions, I would really appreciate the help!
> 
> [crash-2026-01-15_23.51.29-client.txt](https://github.com/user-attachments/files/24995995/crash-2026-01-15_23.51.29-client.txt) [crash-2026-01-15_23.51.27-server.txt](https://github.com/user-attachments/files/24995997/crash-2026-01-15_23.51.27-server.txt) [log.txt](https://github.com/user-attachments/files/24995996/log.txt)

Stalled threads are the bane of my existence as a modpack dev. I'd take a look at any chunk-related optimization mods. That was the main cause in my case.

**Comment by LokeYourC3PH** [comment]
> > Dear community, maybe someone can help or point me in the right direction.
> > I created my own modpack for Minecraft 1.18.2 Forge (partially forked from DawnCraft, but mostly customized for my own needs). We were playing with friends over LAN (without setting up a dedicated server) and ran into a problem where, at some point, the server completely freezes or stops for various reasons:
> > 
> > * a large number of entities on a mob farm
> > * multiple players loading chunks simultaneously in different directions
> > * chests and storage systems at the base
> > * possibly frequent teleportation to each other (sometimes from different dimensions)
> > * or a combination of all of the above
> > 
> > Sometimes, when the server starts freezing, it just lags, then after a while rolls back and continues working. However, in some cases it results in a complete freeze / crash. A crash report is not always generated, but I managed to collect some logs.
> > I tried to analyze the issue using ChatGPT, providing logs and the mod list. It didn’t point to anything definitive, but ModernFix was mentioned as a possible suspect. I haven’t yet started manually disabling mods one by one and testing (since this is time-consuming).
> > ModernFix version: modernfix-forge-5.18.0+mc1.18.2.jar
> > Also mentioned mods (that may cause issue):
> > 
> > * Fast Async World Save
> > * Feature NBT Deadlock Be Gone
> > * ServerCore
> > * It Shall Not Tick
> > 
> > After researching the issue myself, searching for “Server thread stalled / tick timeout” and ModernFix, I came across this GitHub Issues page.
> > Of course, the issue could be caused by another mod or a conflict between multiple mods, but the problems described on this Issues page are very similar to what we experienced during gameplay.
> > P.S. I am attaching the logs I managed to collect while trying to identify the problem. If anyone has ideas or suggestions, I would really appreciate the help!
> > [crash-2026-01-15_23.51.29-client.txt](https://github.com/user-attachments/files/24995995/crash-2026-01-15_23.51.29-client.txt) [crash-2026-01-15_23.51.27-server.txt](https://github.com/user-attachments/files/24995997/crash-2026-01-15_23.51.27-server.txt) [log.txt](https://github.com/user-attachments/files/24995996/log.txt)
> 
> Stalled threads are the bane of my existence as a modpack dev. I'd take a look at any chunk-related optimization mods. That was the main cause in my case.

Throw in Starlight there just for good measures as I mentioned above. I had no idea how much Starlight completely fucks with your game to an extreme extent, since after its removal, my game ran smoother and wasn't as much of a stutter fiesta (I had complete hangups for seconds at a time due to Starlight) anymore. Crazy...

**Comment by CreeperEntertain** [comment]
I'm personally not using Starlight, and wasn't using it at the point of writing my earlier reply.

**Comment by embeddedt** [comment]
The frequency of comments on both #446 and #448 has dropped significantly in recent months. I am suspicious these problems were caused by a bug in another mod which was resolved (or everyone ended up removing it for another reason). I have not been able to find any issue in ModernFix that would explain these crashes. Closing both of these reports for now as they are old and not reproducible (too many mods and other unknowns). If anyone is able to reproduce this with a minimal set of mods as outlined in the current bug report template, please open a new issue.
